pkg/timelock/timelock.go

```diff
@@ -132,63 +132,81 @@ func (tw *Worker) Listen(ctx context.Context) error {
 		return err
 	}
 
+	tw.startLog()
+
+	// Handle OS signals to properly stop/kill the process.
 	stopCh := make(chan string)
-	logCh := make(chan types.Log)
-	sigCh := make(chan os.Signal, 1)
-	signal.Notify(sigCh, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM)
+	go handleOSSignal(stopCh)
 
+	// Update timelock-worker default scheduler delay.
 	tw.updateSchedulerDelay(time.Duration(tw.pollPeriod) * time.Second)
+
+	// Run the scheduler to add/del operations in a thread-safe way.
 	go tw.runScheduler(ctx)
-	tw.startLog()
 
-	// Shutdown gracefully based on OS interrupts.
-	go func() {
-		for {
-			handleOSSignal(<-sigCh, stopCh)
-		}
-	}()
+	// Initialize the subscription.
+	logCh := make(chan types.Log)
+	if err := tw.subscribeAndProcessLogs(ctx, logCh, stopCh); err != nil {
+		tw.logger.Error().Err(err).Msg("failed to subscribe and process logs.")
+		return err
+	}
+
+	defer close(stopCh)
+	defer close(logCh)
```

Comment:
closing log channel away from unsubscribing to the streaming filter query is a bit "dangerous" in that there may be a leftover write from the subscription or the filter query slice.